### ISSUE TITLE
possible for blake3 to have 224 bit?

### ISSUE THREAD

**Issue by gitmko0** [BLAKE3-team/BLAKE3#123]
possible for blake3 to have 224 bit version?

**Comment by gitmko0** [comment]
also a 128bit version?

**Comment by gitmko0** [comment]
128bit for file "hashing" is kind of an overkill "to me". what do u think? coz i would like to give all my image / doc files a unique id name... in fact all data i will eventually store. what's your take on a "shortened" 128bit (16byte) blake3 for this?

advisable? the collision seemed "to me" extremely unlikely.

**Comment by oconnor663** [comment]
You're right that collisions are very unlikely at 128 bits. The math follows the [birthday problem](https://en.wikipedia.org/wiki/Birthday_problem), so you're talking about ~2<sup>64</sup> operations to find a collision. So the questions you have to ask are:

- Does it matter if there's a collision? Could that break something? Use cases like hash tables are designed to expect collisions. Other use cases like cryptocurrencies and TLS rely on collision resistance (or preimage resistance) for security.
- Is the system so large that it's ever going to perform 2<sup>64</sup> hashing operations in practice? Generally no, but it has been done. I think the Google cluster that broke SHA-1 did about that much work.
- Would any adversary ever have a reason to attack the system by forcing it to perform more work (not really feasible at the scales we're talking about) or by performing more work offline themselves (more feasible)?

For a lot of systems, you can confidently answer no to both of those questions. But not all systems, of course. If you're writing library code that might run on other people's systems, it might be reasonable to make conservative design decisions that are safe for everyone. It's also important to think about the cost of 16 extra bytes. Is that totally trivial? Or are you a video game bottlenecking on memory reads? In practice, the biggest tradeoff might be "using cryptographic functions in standard ways that anyone can verify at a glance" vs "doing nonstandard things and then re-checking our assumptions from time to time to make sure we're still safe."

**Comment by xnox** [comment]
> 128bit for file "hashing" is kind of an overkill "to me". what do u think? coz i would like to give all my image / doc files a unique id name... in fact all data i will eventually store. what's your take on a "shortened" 128bit (16byte) blake3 for this?
> 
> advisable? the collision seemed "to me" extremely unlikely.

The same amount of fast expensive computation is performed to calculate the regular or shortened length hash. Storing just the shorter prefix, will reduce the storage requirement of the hash, but not change the time to compute or verify it.

What constraints are causing you to consider computing the full regular size hash, and then throwing half of the result away?

Do you have constraint on storage of hashes? are you hashing many tiny files which are all each smaller than the 256 bits, thus making hashes larger than data itself? If you need to hash and verify a continuous stream, you may want to look into bao for verified streaming.

**Comment by boogerlad** [comment]
@xnox What should be done in the case the input is shorter than the default output of 256 bits? For example, session identifiers are usually 128 bits long from a CSPRNG and stored in a cookie for the client, but on the server side they're stored in the database hashed so that a database leak won't allow an attacker to impersonate a client.

**Comment by boogerlad** [comment]
>  Of course shortening the hash reduces its security guarantees

@oconnor663 the paper mentions that BLAKE3 targets "128-bit security against (second-)preimage, collision, or differentiability attacks. The key length is nevertheless 256 bits, as an extra defense layer against weakly chosen keys, key material leaks, and potentially multi-target attacks." Does this mean that 128 bits security is just a lower bound? Usually cryptographic hash algorithms output size == (second-)preimage security strength. [For example, SHA3-256 has a (second-)preimage resistance of 256 bits](https://en.wikipedia.org/wiki/SHA-3#Instances)

If BLAKE3 output was truncated to 128 bits, how would its security guarantees be reduced specifically? Would everything be cut in half, or just collision resistance?

**Comment by oconnor663** [comment]
> Does this mean that 128 bits security is just a lower bound?

Yes I think that's right. That said, I always defer to @sneves on questions about security claims, because I don't want to get the technical details wrong :)

> If BLAKE3 output was truncated to 128 bits, how would its security guarantees be reduced specifically? Would everything be cut in half, or just collision resistance?

I think everything's going to get cut in half, but like you noted above, first- and second-preimage resistance have 128 bits of margin in practice.

> It doesn't make sense for the output to be more secure than the input, right?

Yes, so for example if all you care about is 128 bits of first- and second-preimage resistance, using a 128 bit hash output is fine. (That's why most MACs are 16 bytes long.) But remember that "I'm giving up collision resistance" is a contract between you and everyone who will ever touch your code from now to the end of time, and if anyone (including future you) forgets the contract, it could turn into a security problem. So make sure you're getting a performance / storage benefit that really matters to you, in exchange for taking on that additional risk / mental effort.

**Comment by sneves** [comment]
Yes, 128 bits are a lower bound, and the only one we commit to. I would expect (2nd)-preimage and so on to be closer to 256 bits, as in the case in previous versions of BLAKE(2), under the assumption the underlying block cipher is well-behaved.
